src/environment.rs

```diff
@@ -0,0 +1,364 @@
+            current: Rc::new(HashMap::new()),
+            previous: RefCell::new(None),
+        }
+    }
+
+    /// Inserts a key-value pair into the Environment.
+    /// If the key was present in the _current_ Environment, the previous value is returned.
+    pub fn insert(&mut self, key: K, value: V) -> Option<V> {
+        if self.was_cloned() {
+            self.current = Rc::new(HashMap::new());
+        }
+        Rc::get_mut(&mut self.current).unwrap().insert(key, value)
+    }
+
+    /// Tries to find the value of a key in the Environment.
+    pub fn get(&self, key: &K) -> Option<V>
+    where
+        V: Clone,
+    {
+        self.iter_layers().find_map(|hmap| hmap.get(key).cloned())
+    }
+
+    /// Creates an iterator that visits all layers from the most recent one to the oldest.
+    /// The element iterator type is `Rc<HashMap<K, V>>`.
+    pub fn iter_layers(&self) -> EnvLayerIter<'_, K, V> {
+        EnvLayerIter {
+            env: if !self.was_cloned() {
+                Some(NonNull::from(self))
+            } else {
+                // if was cloned, current is the same as first of previous (that cannot be empty then)
+                self.previous
+                    .borrow()
+                    .as_ref()
+                    // SAFETY: created from Rc, so cannot be null
+                    .map(|prev| unsafe { NonNull::new_unchecked(Rc::as_ptr(prev) as *mut _) })
+            },
+            _marker: PhantomData,
+        }
+    }
+
+    /// Creates an iterator that visits all elements from the Environment, from the oldest layer
+    /// to the most recent one. It uses this order, so calling `collect` on this iterator to create
+    /// a hashmap would have the same values as the Environment.
+    /// The element iterator type is `(&'env K, &'env V)`, with `'env` being the lifetime of the Environment.
+    pub fn iter_elems(&self) -> EnvElemIter<'_, K, V> {
+        let mut env: Vec<NonNull<HashMap<K, V>>> = self
+            .iter_layers()
+            // SAFETY: all NonNull::new_unchecked comes from pointers created from Rc, so cannot be null
+            .map(|hmap| unsafe { NonNull::new_unchecked(Rc::as_ptr(&hmap) as *mut _) })
+            .collect();
+        // SAFETY: by design, env cannot be empty, and coming from an Rc, it is well aligned and initialized
+        let current_map = unsafe { env.pop().unwrap().as_ref() }.iter();
+        EnvElemIter { env, current_map }
+    }
+
+    /// Creates an iterator that visits all elements from the Environment, from the current layer to the oldest one.
+    /// If values are present multiple times, only the most recent one appears.
+    /// [`iter_elems`] should be preferred, since it does not need to create an intermediary hashmap.
+    /// The element iterator type is `(&'env K, &'env V)`, with `'env` being the lifetime of the Environment.
+    ///
+    /// [`iter_elems`]: Environment::iter_elems
+    ///
+    pub fn iter(&self) -> EnvIter<'_, K, V> {
```

Comment:
Is there a reason to have the `EnvIter` wrapper instead of a directly returning the underlying `hash_map::IntoIter<&'a K, &'a V>` here?

Reply:
It's always about being consistent. All the other kind of iterators have their own structures, so does this one. Ultimately, the compiler should remove it and directly use HashMap::intoiter.

Reply:
Also, `warning: bounds on generic parameters are not enforced in type aliases`